app/common.py

```diff
@@ -0,0 +1,277 @@
+        new file/dir will be populated as the src.
+    2. it deals with the same symlinks by checking the link target,
+        re-generate the symlink if the dst symlink is not the same
+        as the src.
+    3. it will remove files that not presented in the src, and
+        unconditionally override files with same path, ensuring
+        that the dst will be identical with the src.
+    """
+    if not dst.is_dir():
+        raise FileNotFoundError(f"{dst} is not found or not a dir")
+
+    # phase1: populate files to the dst
+    for cur_dir, _, files in os.walk(src, topdown=True, followlinks=False):
+        _cur_dir = Path(cur_dir)
+        _cur_dir_on_dst = dst / _cur_dir.relative_to(src)
+
+        # cover the edge case that dst is not a dir.
+        if not _cur_dir_on_dst.is_dir():
+            _cur_dir_on_dst.unlink(missing_ok=True)
+            _cur_dir_on_dst.mkdir(parents=True)
+            copy_stat(_cur_dir, _cur_dir_on_dst)
+
+        # populate files
+        for fname in files:
+            _src_f = _cur_dir / fname
+            _dst_f = _cur_dir_on_dst / fname
+
+            # src and dst type mismatch, dst is a folder
+            if _dst_f.is_dir():
+                shutil.rmtree(_dst_f, ignore_errors=True)
+
+            # check symlink, re-link if needed
+            if _src_f.is_symlink():
+                dst_symlink_as_src(_src_f, _dst_f)
+                continue
+
+            # finally populate the new file or override the dst file
+            shutil.copy(_src_f, _dst_f, follow_symlinks=False)
+            copy_stat(_src_f, _dst_f)
+
+    # phase2: remove unused files in the dst
+    for cur_dir, dirs, files in os.walk(dst, topdown=True, followlinks=False):
+        _cur_dir_on_dst = Path(cur_dir)
+        _cur_dir_on_src = src / _cur_dir_on_dst.relative_to(dst)
+
+        # remove unused dir
+        if not _cur_dir_on_src.is_dir():
+            shutil.rmtree(_cur_dir_on_dst, ignore_errors=True)
+            dirs.clear()  # stop iterate the subfolders of this dir
+            continue
+
+        for fname in files:
+            _src_f = _cur_dir_on_src / fname
+            if not (_src_f.is_file() or _src_f.is_symlink()):
+                (_cur_dir_on_dst / fname).unlink(missing_ok=True)
+
+
+class SimpleTasksTracker:
+    def __init__(
+        self, *, semaphore: Semaphore, title: str = "simple_tasks_tracker"
+    ) -> None:
+        self.title = title
+        self._wait_interval = cfg.STATS_COLLECT_INTERVAL
+        self.last_error = None
+        self._se = semaphore
+
+        self._interrupted = Event()
+        self._register_finished = False
+
+        self._in_counter = itertools.count()
+        self._done_counter = itertools.count()
+        self._in_num = 0
+        self._done_num = 0
+
+    def register(self):
+        if self._interrupted.is_set() or self._register_finished:
+            return
+
+        self._se.acquire()
+        self._in_num = next(self._in_counter)
```

Comment:
self._in_counter is not thread-safe(at least the document doesn't say that) and the semaphore value is not 1, so I think updating `_in_num` is not protected. `_done_num` is also the same.

Reply:
Could add the comment above?

Reply:
updated in [0024c13](https://github.com/tier4/ota-client/pull/136/commits/0024c13369832d705cc6ddc701d5addc5506c48f)